Feature: tdigest aggregate

thanodnl opened the pull request:
DESCRIPTION: Adds support to partially push down tdigest aggregates

tdigest extensions: https://github.com/tvondra/tdigest
~~when testing use; https://github.com/citusdata/tdigest/tree/fix/in-out
This branch contains a patch for parsing a `tdigest` from a `text` which is required for Citus to fetch partial results from the workers.~~ patch has been merged upstream

This PR implements the partial pushdown of tdigest calculations when possible. The extension adds a tdigest type which can be combined into the same structure. There are several aggregate functions that can be used to get;
 - a quantile
 - a list of quantiles
 - the quantile of a hypothetical value
 - a list of quantiles for a list of hypothetical values

These function can work both on values or tdigest types.

Since we can create tdigest values either by combining them, or based on a group of values we can rewrite the aggregates in such a way that most of the computation gets delegated to the compute on the shards. This both speeds up the percentile calculations because the values don't have to be sorted while at the same time making the transfer size from the shards to the coordinator significantly less.

TODO:
 - [x] Create patch on top of tdigest extension to correctly serialize/deserialize tdigest type to text.

marcocitus commented on src/backend/distributed/planner/multi_logical_optimizer.c:
```diff
@@ -3178,6 +3370,66 @@ GetAggregateType(Aggref *aggregateExpression)
+	 * hit.
+	 */
+	if (StartsWith(aggregateProcName, "tdigest"))
+	{
```
Given that this condition will only be true in very specific cases for queries that are low throughput / execution-heavy, do we really need OID caching for the TDigest functions? 

We mainly do caching because some functions OIDs are checked in practically every query (e.g. citus_extradata_container, read_intermediate_results), but we don't do it for HLL functions for instance.

What if we just do LookupFuncName in the TDigest functions and don't worry about cache invalidation? There's actually another layer of caching in PostgreSQL when looking up functions by name, so it will still be pretty fast.

It would also be nice to move the TDigest functions out of metadata_cache.c, which is already enormous.

thanodnl commented:
@marcocitus For rollups you can use the following two aggregates:
 - `tdigest(value, compression)`: This aggregates is what produces a `tdigest` value. This value can be stored in a table - I actually added a test for this to the upstream project.
 - `tdigest(tdigest)`: This aggregate can combine multiple stored `tdigest` values into 1, this is the union you are looking for I guess.

Lastly the `tdigest_percentile(_of)` aggregates (these are not extract functions like `hll_cardinality`) could take multiple `tdigest` values already, and combine them before returning the percentiles. They will perform an implicit union if more than 1 `tdigest` value gets combined in the aggregate group.

Contrary to the HLL and TopN extensions which have explicitly union in their names it uses overloading of aggregates to provide the union functionality. And the functions to read percentiles from the `tdigest` values are not functions but aggregates instead; which makes it even more natural to combine, but might be slightly harder to understand if you are used to `agg` -> `unionagg` -> `extract function`.

Hope this explains how to union.

serprex commented on src/backend/distributed/metadata/metadata_cache.c:
```diff
@@ -157,6 +157,20 @@ typedef struct MetadataCacheData
 	char databaseName[NAMEDATALEN];
+
+	/* cached lookups for the tdigest extension */
+	Oid tdigestExtensionSchema;
```
So there's caching issues, & HLL/topn isn't in here, what's the cost of not caching this?

thanodnl replied:
The cost is that it will need to perform up to 10 lookups into the catalog (2x once for the extension schema and once for the function/type oid). To come to a conclusion which aggregate we are targeting.
Secondly each worker and coordinator part of the query requires ~3 lookups each to construct the query part.

coming to up to 26 catalog lookups.
Luckily this only happens if the aggregate starts with `tdigest`, so it will only perform them if there is a high likely hood we are actually querying for a tdigest.

Given the execution time is already high on queries where you want to use tdigest these lookups during planning probably are not significant to the execution of the query. Hence Marco already suggesting to not cache them.

I am moving these functions to their own file to not pollute the metadata cache file more.

serprex commented on src/backend/distributed/metadata/metadata_cache.c:
```diff
@@ -4222,3 +4236,366 @@ poolinfo_valid(PG_FUNCTION_ARGS)
+
+
+/*
+ * TDigestExtensionTypeOid preforms a lookup for the Oid of the type representing the
```
> preforms

performs

serprex commented on src/backend/distributed/metadata/metadata_cache.c:
```diff
@@ -4222,3 +4236,366 @@ poolinfo_valid(PG_FUNCTION_ARGS)
+ * If the aggregate is not found InvalidOid is returned.
+ */
+Oid
+TDigestExtensionAggTDigestPercentileOf3a(void)
```
Feel all these functions could be made much shorter with a helper function

thanodnl replied:
Possibly. Now that we move away from the cached version I might make a helper function where we solely pass the signature of the function we are looking for. That should make these functions pretty compact aliases to the function signature.

thanodnl commented:
I have moved the tdigest functions to their own files.

Removed the caching, the lookups are guarded by a simple check where it only iterates through catalogs if there is a high chance of finding a match. Lookups are redundant, especially to the schema. However optimising it might not be worth it at the moment as it is easy to cause poisoned caches with moving the extension around or removing and reinstalling it.

Also incorporated the idea of @serprex to make the lookups less redundant and easier to map to the functions we are searching for.

thanodnl commented:
sidenote; once this change is approved I will fix the merge conflict the branch is having with enterprise, before merging into community.

serprex commented on src/backend/distributed/planner/multi_logical_optimizer.c:
```diff
@@ -3178,6 +3371,66 @@ GetAggregateType(Aggref *aggregateExpression)
+	 */
+	if (StartsWith(aggregateProcName, "tdigest"))
+	{
+		if (aggFunctionId == TDigestExtensionAggTDigest1())
```
Why check each OID rather than having something like `AggregateNames` to check `aggregateProcName` against?

thanodnl replied:
the name of this aggergate is `tdigest` which collides between
`tdigest(tdigest)` and `tdigest(value, compression)`. Both will have `tdigest` as their `aggregateProcName`.

Same goes for `tdigest_percentile` (4 overloads) and `tdigest_percentile_of` (4 overloads).

To make sure we can distinguish between them it is easiest to look them up via their Oid.
